### DISCUSSION TITLE
How do I publish a package to Pulsar

### DISCUSSION THREAD

**Discussion by manngo** [pulsar-edit/.github#91] [Community Packages]
I have an existing package written for Atom. Of course, I want to complete my migration, but I can’t find any documentation.

In the past, I ran something like:

```sh
apm publish minor
```

Of course, there isn’t any `apm` now, especially on my new Mac, which never had Atom installed. A passing reference in https://github.com/pulsar-edit/.github/discussions/63 suggests it is:

```sh
pulsar -p
```

but I can’t find a `pulsar` command either.

What is the missing step?

**Comment by Daeraxa** [comment]
The documentation is on the website - https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#publishing

However, as of right now the authentication hasn't been made live to allow publishing just yet (coming very soon though). 

The missing `pulsar` command will be because it is missing from your PATH. What OS and release type are you using?
EDIT - Seems you are on mac after re-reading. The missing PATH install is a known issue however a user did manage to solve this themselves - see this thread on [Discord](https://discord.com/channels/992103415163396136/992103415163396139/1052302826270887945). Text below:

> With $ which atom, I found that the alias was in /usr/local/bin/atom, so I changed /Applications/Atom.app/Contents/Resources/app/atom.sh to /Applications/Pulsar.app/ Contents/Resources/app/pulsar.sh copy and replaced Atom with Pulsar.Perhaps there is a smarter method, but thanks for the your guidance.

**Reply by manngo** [reply]
While cruising around I did find `/Applications/Pulsar.app/Contents/Resources/pulsar.sh`. I wasn’t sure whether this was it, and there’s no reference to packages.

Would I run:

```sh
/Applications/Pulsar.app/Contents/Resources/pulsar.sh -p publish minor
```

… ?

**Reply by manngo** [reply]
BTW How do I get access to the Discord channel you linked to?

**Reply by confused-Techie** [reply]
@manngo to join the Discord, click the link in our Org readme, which I'll go ahead and [link](https://github.com/pulsar-edit#pulsar-a-community-led-hyper-hackable-text-editor) to.

As for the command needed, that likely would work. But depending on your OS it is a known issue that Pulsar doesn't put itself on the path as we would expect. But if you'd like you can stay up to date on this bugs status [on this issue](https://github.com/pulsar-edit/pulsar/issues/187)

**Reply by anonCoffee** [reply]
@confused-Techie Discord is also just another proprietary headache and nothing else.

**Reply by Daeraxa** [reply]
> @confused-Techie Discord is also just another proprietary headache and nothing else.

Noted and ignored

**Reply by anonCoffee** [reply]
Ignored for bad. Do whatever.

**Reply by Daeraxa** [reply]
Just to add to this, we are looking at making a Matrix space with a discord bridge for those who would rather not use it.

**Reply by confused-Techie** [reply]
Yes, and also if you'd like all communication can happen on GitHub, which I know you are also not a fan of. But then you could communicate strictly on Mastodon. We try to have options out there for everyone

**Comment by confused-Techie** [comment]
I'd like to add @manngo if your package was previously published to Atom, it's very likely that it's already on the Pulsar Package Repository.

Is it possible this is your package?

https://web.pulsar-edit.dev/packages/web-developer-tools

**Reply by manngo** [reply]
Thanks, I saw that. I was wondering about publishing my updates (I noticed a new bug, caused, I think, by an change in file handling).

**Reply by confused-Techie** [reply]
> Thanks, I saw that. I was wondering about publishing my updates (I noticed a new bug, caused, I think, by an change in file handling).

Unfortunately like @Daeraxa mentioned currently Authentication is disabled on the backend, meaning no one can login yet to update packages, or publish new ones.

Although in the very near future we are planning to enable authentication and at that time you should be able to publish an update with no issues after creating a Pulsar User Account.

If everything goes to plan to the backend will automatically recognize you as the owner of the package, as long as you still have write access to the repo.

**Reply by manngo** [reply]
Meanwhile, if I publish to Atom with `atm` would it find its way to Pulsar?

**Reply by mauricioszabo** [reply]
APM and all Atom servers are offline now, so... unfortunately, no.

Even if they where online, Atom was plagued with SPAM packages in the last few months, so we could not synchronize anymore...

**Reply by manngo** [reply]
Thanks. I’ll wait until then.

Thanks for all of your hard work and help.

**Reply by confused-Techie** [reply]
> Thanks. I’ll wait until then.
> 
> Thanks for all of your hard work and help.

Actually Authentication is now on! Take a look at the [release](https://github.com/pulsar-edit/package-backend/releases/tag/v1.0.0), which has some links to docs to help you get started and make an account

**Comment by manngo** [comment]
OK, I am still trying to get this to work for me. Here is what I have attempted:

-	First, I have logged in to https://api.pulsar-edit.dev/api/login
-	Second, I went into my project directory and attempted to publish it according to https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#using-ppm-pulsar-package-manager:

	```sh
	cd github/atom-web-tools
	/Applications/Pulsar.app/Contents/Resources/pulsar.sh -p publish minor
	```

I am getting the error message:

> The application //Applications cannot be opened for an unexpected reason, error=Error Domain=NSOSStatusErrorDomain Code=-10827 "kLSNoExecutableErr: The executable is missing" UserInfo={_LSLine=4101, _LSFunction=_LSOpenStuffCallLocal}

So obviously I’m still missing something.

I am guessing that the other requirements are still OK, as the package has been published and updated a number of times.

**Reply by confused-Techie** [reply]
Hopefully someone from @pulsar-edit/core can chime in on the error you are seeing.

As for the backend and publishing looks like you're on the right track, I'd only make sure that you use your new API token to login to PPM.

`pulsar -p --login PULSAR_API_TOKEN` (I think is the command) Then you should be able to publish with that command.

I'd also just like to advise making sure you are on the most up to date Alpha build, as we have been sorting a lot of PPM issues recently so that will give you the best chances to publish successfully

**Reply by manngo** [reply]
Thanks for the login step: I didn’t know how to use it.

However, I’m still getting the same error message.

I’ll check on the Pulsar version.

**Reply by Daeraxa** [reply]
> cd github/atom-web-tools
/Applications/Pulsar.app/Contents/Resources/pulsar.sh -p publish minor

You aren't trying to execute pulsar.sh are you? That isn't the pulsar executable, that is a supporting shell script.

**Reply by manngo** [reply]
No, that’s what I used at first, but I have discovered that it’s `/Applications/Pulsar.app/Contents/MacOS/Pulsar` . I don’t know why I didn’t see that executable before.

**Comment by manngo** [comment]
I am getting one step closer, I think.

The steps now are:

```sh
$	cd github/atom-web-tools
$	/Applications/Pulsar.app/Contents/MacOS/Pulsar -p login
Token>	…
$	/Applications/Pulsar.app/Contents/MacOS/Pulsar -p publish minor
```

However, although the `login` step above says that the token is saved to the Keychain, the `publish` step still wants a username and password, and then tells me that password authentication was removed on August 13, 2021.

So I’m getting closer. I think.

**Reply by confused-Techie** [reply]
Okay so that's good.

Essentially what you are seeing here is part of a bigger discussion we are having, but I'll try to give you the TLDR.

Essentially, historically your API Token for Atom also allowed access to your GitHub account, so APM would access your GitHub account, tag the release and commit as needed for it's purposes, then tell the Atom API to publish the new package, then it would publish a new version.

But now, since we aren't GitHub that's changed. Essentially your Pulsar API Token is just that, a token to access the Pulsar API. PPM as it's mostly still APM tries to tag and publish to your GitHub account. So `--login` provides PPM with your Pulsar API Token, but PPM still needs your GitHub Username and Password. (Which GitHub doesn't support just logging in with a password through the CLI so you need to make a PAT token, and login here with your GitHub Username and PAT token).

So now PPM has your Pulsar API Token, and the credentials to access your GitHub account.

Once it has this it will now be able to Tag and commit to your GitHub account like it tries to, and will be able to publish the package to the new backend. 

So with that you will be golden!

But fair warning, remember how I said APM tries to publish a package, then publish a new version? It will still do this but the backend now will automatically publish the version, so PPM will fail there. So will still return an error. 

What this means in all reality, what success will look like for you right now is about this:

```
$ ~/pulsar/pulsar -p publish -t v0.2.3
Registering atom-developer-tools ✓
Preparing and tagging a new version ✓
Pushing v0.2.3 tag ✓
Publishing atom-developer-tools@v0.2.3 ✗
Creating new version failed: Application Error
```

Obviously slightly different, but you get the drift, if it says publishing and returns an error go ahead and check what the latest version the backend has [here](https://web.pulsar-edit.dev/packages/web-developer-tools).

**Comment by manngo** [comment]
Here are my revised steps:

```
#	Package Directory
	cd github/package-diretory

#	Log in to Pulsar
	/Applications/Pulsar.app/Contents/MacOS/Pulsar -p login
	Token>	…

#	Git Hub Personal Access Token
	See https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
	
#	Publish
	/Applications/Pulsar.app/Contents/MacOS/Pulsar -p publish -t x.x.x
	User:		[user]
	Password:	[Access Token]
```

It appears to have worked, but I can’t update the package on Pulsar. When I try, I get the following error:

> Updating to “web-developer-tools@0.4.4” failed.Hide output…
> 
> npm ERR! code E404
> npm ERR! 404 Not Found - GET https://api.pulsar-edit.dev/api/packages/web-developer-tools/versions/0.4.4/undefined
> npm ERR! 404 
> npm ERR! 404  'https://api.pulsar-edit.dev/api/packages/web-developer-tools/versions/0.4.4/tarball' is not in the npm registry.
> npm ERR! 404 Your package name is not valid, because 
> npm ERR! 404  1. name can only contain URL-friendly characters
> npm ERR! 404 
> npm ERR! 404 Note that you can also install from a
> npm ERR! 404 tarball, folder, http url, or git url.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/mark/.pulsar/.apm/_logs/2022-12-28T03_54_32_795Z-debug.log

I don’t know whether that’s in the right direction … ?

**Reply by confused-Techie** [reply]
I'll be honest, I'm rather confused at these logs. The error you are getting doesn't seem to exist within PPM, the Backend, or within Pulsar.

Looks like it's attempting to install this package from NPM?

Anyone from @pulsar-edit/core have any ideas?

**Reply by DeeDeeG** [reply]
That error message comes out of the underlying `npm` package used in `ppm`, apparently.

https://github.com/npm/cli/blob/v6.14.18/lib/utils/error-message.js#L196-L225

It's for when the package you are publishing is not found at the registry you're trying to publish to.

**Reply by DeeDeeG** [reply]
> 404 Not Found - GET https://api.pulsar-edit.dev/api/packages/web-developer-tools/versions/0.4.4/undefined

That `undefined` at the end of that URL looks at least mildly suspicious. Perhaps we have some sort of error in `ppm` or the backend that's not forming that URL quite correctly? Or maybe some garbled data was fed through `ppm` or the backend to cause that `undefined` to come out the other end?

(I do hope to take a deep dive through the publishing process myself soon, and debug/troubleshoot some of these things, but no promises as to when I will next have the time to do so.)

---

\[EDIT: Okay, so the [tarball URL](https://api.pulsar-edit.dev/api/packages/web-developer-tools/versions/0.4.4/tarball) for this 0.4.4 package version (URL ending in `/tarball`) redirects to the same URL but ending in `/undefined` in place of `/tarball`, for some reason.

Maybe the tarball was expected by our backend, but was never uploaded (or the upload didn't finish...)? I admit this is just a guess as to why there is the redirect to `/undefined`. Something is wrong with what we have on our end for this package version, but it's unclear to me whether or not it originated/transmitted wrong from `ppm`.\]

---

EDIT again: I notice this package's newer versions' API JSON responses don't have a `tarball_url` field, which should be pointing to GitHub.com. They only have `dist.tarball`, but older versions had both `tarball_url` and `dist.tarball`.

---

EDIT to add: https://web.pulsar-edit.dev/packages/web-developer-tools gives me a "This Page enccounted an Error! Sorry!" error page, and strangely enough, my browser dev tools show a 505 error.

<details><summary>Tangent/Off-topic: About the 505 error status code (click to expand):</summary>
"505 HTTP Version Not Supported

The HyperText Transfer Protocol (HTTP) 505 HTTP Version Not Supported response status code indicates that the HTTP version used in the request is not supported by the server. "

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/505
</details>

<details><summary>Tangent: why a 505 error code? (click to expand):</summary>

(I am assuming it's just throwing a too-specific "505" error, but that my HTTP version isn't the real problem? Should be a generic 500 internal server error, maybe? But if part of the code hard-codes 505 by mistake, maybe that's a good thing to do a code search for? Where in the code does it throw a 505 error? EDIT: Yeah, looks like it's used by the frontend repo for any non-404 error status: https://github.com/pulsar-edit/package-frontend/blob/66a4742a3b38176cb74bce53cc7689fee42b5143/src/utils.js#L18-L30)

</details>

... Whereas for example the `github` package page loads fine for me: https://web.pulsar-edit.dev/packages/github

**Comment by manngo** [comment]
I might add that updating `path-copy` from `0.14.0` to `0.17.0` was successful. Thing might suggest that updating works, but my on publishing didn’t go through.

This is both on my current M1 Mac and and an older Intel iMac.

**Comment by cschindl** [comment]
Apparently the console command does not work under windows. When I execute it with `pulsar -p help` or `pulsar -p login` nothing happens afterwards. 

I am therefore unfortunately not able to execute the login or publish command.

So I tried to do it with apm (Pulsar Package Manager) like I did with atom. Here I could add the token and also publish.

The repository was updated on github and a new tag was added. Unfortunately there was no new release and nothing changed on https://web.pulsar-edit.dev/packages/ftp-remote-edit. I have therefore removed the commits and the tag again.

What do I have to do if I want to update the package?

**Reply by topa-LE** [reply]
Would be really nice if you could install the update without errors.

![img_158](https://user-images.githubusercontent.com/63888024/226121175-ccd92758-e435-4bb9-857b-1e4df70035f9.png)

**Reply by confused-Techie** [reply]
Did you actually use `apm` or `ppm`?
If it was `apm` that still would point to Atoms package registry, which no longer exists. So it would've only published the GitHub tag.

If it was done with `ppm` that likely worked, but there is currently [this](https://github.com/pulsar-edit/package-backend/issues/97) bug which could result in the version not being shown directly within settings or the website, but still installable as the latest version.

For example, I think this bug is what you saw with `ftp-remote-edit`. While the website still shows the current latest version as `0.18.2` when going to install the latest package version I get `0.20.0`. So it looks like it was published to the server

**Reply by cschindl** [reply]
I used the Pulsar Package Manager. The executable `apm.cmd` is located in the folder `C:\Program Files\Pulsar\resources\app\ppm\bin` on Windows. I was therefore a bit confused by the name of the executable. However, running the apm.cmd clearly opens the Pulsar Package Manager.

And you are right. The registry contained the new version `v0.20.0`, so the update worked. I was just surprised that it was not displayed on the website in the new version.

After I added the `v0.20.0` tag, the package could also be updated via Pulsar without any problems.

Thanks for the quick help.

P.S. It would be great if the documentation for publishing a package could be updated. The information at: [https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#publish-your-package](https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#publish-your-package) unfortunately contains a lot of misinformation and also the previous login is not described there.

**Reply by confused-Techie** [reply]
@cschindl Okay I see what you mean now. But you are right about the docs needing an update. There is some work to do a large refactor on those as we want to move away from the Atom style Launch Manual anyway.

But also the bug I mentioned earlier I was able to fix [today](https://github.com/pulsar-edit/package-backend/pull/126) and now if you'd like publishing another version will get the website updated as well. So those should no longer be an issue. But I'm glad you got back to me and let me know what's going on, thanks a ton!

**Comment by manngo** [comment]
OK, things have moved on since I first ask this question.

I would now like to try updating the package again. Is this the correct process?

```bash
cd /path/to/package
/Applications/Pulsar.app/Contents/Resources/pulsar.sh -p publish patch # | major | minor
```

If so, I’m getting a number of different errors:

> remote: Invalid username or password.
> remote: Support for password authentication was removed on August 13, 2021
> fatal: Authentication failed for 'https://github.com/manngo/atom-web-tools/'

I have already applied the token.

BTW thanks to @confused-Techie for looking at my other issue https://github.com/pulsar-edit/package-backend/issues/125.

**Reply by confused-Techie** [reply]
That is the correct process.

But things are a little tricky, since the token to the API does not authenticate you to GitHub, which PPM needs to push the new tag for your package to github. You'll need to provide a token for PPM to access GitHub, as well as the Pulsar API.

We've touched on this before, on this [earlier conversation](https://github.com/orgs/pulsar-edit/discussions/91#discussioncomment-4505820) on this thread.

**Reply by manngo** [reply]
> That is the correct process.
> 
> But things are a little tricky, since the token to the API does not authenticate you to GitHub, which PPM needs to push the new tag for your package to github. You'll need to provide a token for PPM to access GitHub, as well as the Pulsar API.
> 
> We've touched on this before, on this [earlier conversation](https://github.com/orgs/pulsar-edit/discussions/91#discussioncomment-4505820) on this thread.

This is where I go around in circles:

```bash
/Applications/Pulsar.app/Contents/MacOS/Pulsar -p login
> [Information]
> Press [Enter] to open your account page.
> Token> …
> Saving token to Keychain ✓

/Applications/Pulsar.app/Contents/Resources/pulsar.sh -p publish patch
> Preparing and tagging a new version ✓
> Pushing v0.5.2 tag Username for 'https://github.com': manngo
> Password for 'https://manngo@github.com': …

> remote: Support for password authentication was removed on August 13, 2021.
```

I tried using both my github password and the access token.

**Reply by confused-Techie** [reply]
Are you using the same access token for the Pulsar API when it asks for your password?

So I know it's a little funky. We are trying to get everything updated and more in sync. But let me lay it out a bit.

PPM tries to do two things, publish to the Pulsar Package Registry, and push a new tag and commit to your GitHub repo.

#### Pulsar Package Registry 

This needs your API key to access Pulsar. The one you got when making your Pulsar Account. This, while yes it is a GitHub token, CANNOT be used to login to your GitHub account. The token you are given for Pulsar by default does not have any permissions to write to your GitHub at all. The only thing this token can do is check if you can write to a repo. That's because that's all the backend needs.

#### GitHub

But when PPM interacts with GitHub it needs a token that actually has write access to your GitHub account. Again not the one for your Pulsar Account.

So this means that running `pulsar -p --login PULSAR_API_TOKEN` gives PPM access to the Pulsar Package Registry. 
Then when you go to publish a package and it asks for your username that's your GitHub username, and it asks for your password, you need to generate a different GitHub PAT that has access to write to your GitHub account.

**Reply by manngo** [reply]
Ah, of course.

I have now got as far as the next problem.

I have applied the pulsar access token. Then when it comes to authenticating I apply my GitHub access token, which is, of course different.

I now get:

> Creating new version failed: Application Error

Sorry to be a pain …

**Reply by confused-Techie** [reply]
No worries @manngo, but weird it doesn't give a more specific error message.

Weird thing is you've published a few updates, is this for a different package? Or the same `web-developer-tools`

Also you could check the website to see if it's picking up the new changes

**Reply by manngo** [reply]
This is the only package I have.

The website still says 0.4.0.

**Reply by confused-Techie** [reply]
So actually taking a check at the logs I'm seeing:

```
[TRACE]:: manngo Attempting to publish a new package version 
[TRACE]:: User Failed Git Ownership Check: Server Error
...[REDACTED]... "POST /api/packages/web-developer-tools/versions?tag=v0.5.6 https" 500 912ms
```

And as far as I can tell the only instances when this error would occur would be:
* The request to github had failed when attempting to determine if you own the package. This failure could be a 505 or 404, really anything other than a 200.
* The credentials supplied resulted in GitHub returning a 401 status code, which is still returning a Server Error rather than bad auth, since I haven't gotten to refactoring error messages in that file yet.

So what I'd recommend, is trying again, in case it is because of a temporary issue of the backend contacting GitHub, if you get the same error message then what you could do is attempt to use something like Postman and hit `https://api.github.com/repos/manngo/atom-web-tools/collaborators?page=1` where you set `Authorization` to be `Bearer <Your Pulsar API Token here>`.

If that fails then that's the root cause, if it succeeds though would you mind sharing the output? It's possible the backend might be checking incorrectly or GitHub has changed the result of that hit.

I'm still investigating now to try and see where the root cause is

**Reply by manngo** [reply]
The issue may bet temporary, but so far I’m still getting the error.

I’m not familiar with Postman. Can you point me in the right direction?

**Reply by manngo** [reply]
Using Curl as per https://docs.github.com/en/rest/collaborators/collaborators#list-repository-collaborators

```bash
curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer …"\
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/manngo/atom-web-tools/collaborators?page=1
```

I got:

> no matches found: https://api.github.com/repos/manngo/atom-web-tools/collaborators?page=1

I don’t know whether I’m completely off the mark

**Reply by confused-Techie** [reply]
Ah that works to, probably should've recommended Curl. Thanks @manngo 

But super interesting it's saying `no matches` found. I have a feeling that's because of the question mark? Some CLIs don't parse it properly, maybe try excluding the page all-together or otherwise wrap the URL in quotes

**Reply by manngo** [reply]
I tried without the `?page=1` and got:

```
{
  "message": "Must have push access to view repository collaborators.",
  "documentation_url": "https://docs.github.com/rest/collaborators/collaborators#list-repository-collaborators"
}
```

Is that a good or bad sign?

**Reply by confused-Techie** [reply]
Well @manngo, if you are using your Pulsar API key as `Authorization: Bearer <PULSAR_API_KEY>` then no, but explains our issue.

The token being used isn't able to access your repo. It could mean your API token has expired. In which case if you go to the Pulsar Website and log back in (you may need to logout first) then you can refresh this token automatically and use the new one.

So I'd just visit [the Pulsar Frontend](https://web.pulsar-edit.dev) and sign back into your account, and take the new API key given there.

**Reply by manngo** [reply]
I’ve just got a new Pulsar key and go the same result.

**Reply by confused-Techie** [reply]
So even with a brand new Pulsar API key you can't access this API Endpoint? Since if not that does explain why Pulsar can't publish.
Since GitHub is saying that token doesn't provide push access to this repo.

I'm somewhat at a loss, because this would be handled all on GitHub's end. Is it possible that there's some sort of security setting enabled on the repo? But I'll try doing some research since I've never seen this before.

**Reply by manngo** [reply]
Would it make any sense to delete this and start again?

**Reply by confused-Techie** [reply]
I wouldn't think so. And the issue isn't with the Pulsar Package, it'd possible be from GitHub. Since we need to figure out why GitHub is denying the token. One thing you could do is use the token you generated earlier to use as a password with PPM, and use that token for the request to the contributor api endpoint instead. That may tell us if the Pulsar API key doesn't have the right permissions to your account

**Reply by confused-Techie** [reply]
@manngo Maybe keep an eye out here, since I'm really not sure what issue it is that is only effecting you (as far as I know)

But I've just added a [PR](https://github.com/pulsar-edit/package-backend/pull/145) to add new data to errors that are returned when attempting to publish a new version. So if you'd like to wait until this PR is merged then try again, we can hopefully see what exactly is going wrong.

**Reply by manngo** [reply]
Thanks. Will that be in version 104?

**Reply by confused-Techie** [reply]
@manngo No, changes to the backend are live as soon as they are published, they don't have to worry about Pulsar versions. So once that PR is merged (give or take a few hours) the changes will be live and you can test it again

**Reply by manngo** [reply]
Sorry for the time gap.

I have tried again, and I’m getting the same issues. I removed old tags, since they were getting out of control, and tried to republish. I still get:

> Creating new version failed: Application Error

Using `curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer …" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/repos/manngo/atom-web-tools/collaborators` I get:

```
https://api.github.com/repos/manngo/atom-web-tools/collaborators
{
	"message": "Requires authentication",
	"documentation_url": "https://docs.github.com/rest/collaborators/collaborators#list-repository-collaborators"
}
```

Again I can’t use `?page=1` on the end.

I have already gone through the `/Applications/Pulsar.app/Contents/MacOS/Pulsar -p login` process.

**Reply by confused-Techie** [reply]
You're okay @manngo. So yeah if that API call to GitHub is still returning an error, no matter what you do to Pulsar or PPM, publishing will fail. The backend relies on the results of this API call to publish a package, or really it relies on this API call to ensure you own the package, and should be able to publish it.

I'm really not quite sure what's causing this API call to fail, but it'd either be an issue with your token, or an issue with the repo, or an issue with GitHub itself.

I'd recommend seeing if you can create a brand new API token from the GitHub UI itself, and us that token for this API call. If it succeeds it can at least tell us that your Pulsar API token doesn't have the permissions needed. And once we know what permissions work, we can determine what's different between the two.

**Reply by cschindl** [reply]
I tried to update my package today. Unfortunately, I received the following error message:

```
PS C:\Users\Carsten\.pulsar\packages\ftp-remote-edit> ppm publish minor
Preparing and tagging a new version done
Pushing v0.21.0 tag done
Publishing ftp-remote-edit@v0.21.0 failed
Internal Server Error
```

The new tag has been set and also correctly pushed to the master on github. What do I have to do to successfully update the package?

**Reply by confused-Techie** [reply]
@cschindl I took a quick look at this, and while everything on PPM is succeeding, I was able to find an issue on the backend when attempting to publish the package `ftp-remote-edit`.

The backend returned the following error messages:

```
[WARNING]:: auth.verifyAuth() API Call returned: 401
[TRACE]:: auth.verifyAuth() API Call Returning Bad Auth
```

While the backend should've reported this too you, it looks like this [PR](https://github.com/pulsar-edit/package-backend/pull/247) which I merged just the other day encountered an error trying to ensure to scrub any sensitive details from the actual error message that you ended up receiving such a generic error message.

---

To solve this, I'd recommend checking the following:

* That you have provided PPM with your Pulsar Package Registry API Token
* That you have at least write permissions on the remote repository you are trying to publish.

If you feel like the authentication failure is still an error of the backend, if you could provide the actual repository you are trying to publish it'd be appreciated. Thanks for reporting this!

**Reply by DeeDeeG** [reply]
It seems possible the token was set to expire after some time, and has expired since the last attempt to publish a package? Worth checking whether the token is now expired.
